src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/TargetRecordingOptionsPatchHandler.java

```diff
@@ -0,0 +1,125 @@
+ * SOFTWARE.
+ * #L%
+ */
+package com.redhat.rhjmc.containerjfr.net.web.handlers;
+
+import java.util.Arrays;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.openjdk.jmc.flightrecorder.configuration.recording.RecordingOptionsBuilder;
+
+import com.google.gson.Gson;
+
+import com.redhat.rhjmc.containerjfr.commands.internal.RecordingOptionsBuilderFactory;
+import com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer;
+import com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer.OptionKey;
+import com.redhat.rhjmc.containerjfr.net.AuthManager;
+import com.redhat.rhjmc.containerjfr.net.TargetConnectionManager;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.ext.web.RoutingContext;
+
+class TargetRecordingOptionsPatchHandler extends AbstractAuthenticatedRequestHandler {
+
+    static final String PATH = TargetRecordingOptionsGetHandler.PATH;
+    private final RecordingOptionsCustomizer customizer;
+    private final TargetConnectionManager connectionManager;
+    private final RecordingOptionsBuilderFactory recordingOptionsBuilderFactory;
+    private final Gson gson;
+
+    @Inject
+    TargetRecordingOptionsPatchHandler(
+            AuthManager auth,
+            RecordingOptionsCustomizer customizer,
+            TargetConnectionManager connectionManager,
+            RecordingOptionsBuilderFactory recordingOptionsBuilderFactory,
+            Gson gson) {
+        super(auth);
+        this.customizer = customizer;
+        this.connectionManager = connectionManager;
+        this.recordingOptionsBuilderFactory = recordingOptionsBuilderFactory;
+        this.gson = gson;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.PATCH;
+    }
+
+    @Override
+    public String path() {
+        return PATH;
+    }
+
+    @Override
+    void handleAuthenticated(RoutingContext ctx) throws Exception {
+        Map<String, String> updatedMap =
+                connectionManager.executeConnectedTask(
+                        getConnectionDescriptorFromContext(ctx),
+                        connection -> {
+                            MultiMap attrs = ctx.request().formAttributes();
+                            Arrays.asList("toDisk", "maxAge", "maxSize")
+                                    .forEach(
+                                            key -> {
+                                                if (attrs.contains(key)) {
+                                                    OptionKey.fromOptionName(key)
+                                                            .ifPresent(
+                                                                    optionKey ->
+                                                                            customizer.set(
+                                                                                    optionKey,
+                                                                                    attrs.get(
```

Comment:
Last thing:

```
$ curl -vk -X PATCH -d "maxAge=no&toDisk=123" -H "X-JMX-Authorization: Basic $(echo -n user:pass | base64)" https://0.0.0.0:8181/api/v1/targets/localhost/recordingOptions
*   Trying 0.0.0.0:8181...
* TCP_NODELAY set
* Connected to 0.0.0.0 (127.0.0.1) port 8181 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=US; O=Red Hat; CN=container-jfr
*  start date: Sep  9 16:27:38 2020 GMT
*  expire date: Mar  8 16:27:38 2021 GMT
*  issuer: C=US; O=Red Hat; CN=container-jfr
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> PATCH /api/v1/targets/localhost/recordingOptions HTTP/1.1
> Host: 0.0.0.0:8181
> User-Agent: curl/7.66.0
> Accept: */*
> X-JMX-Authorization: Basic dXNlcjpwYXNz
> Content-Length: 20
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 20 out of 20 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-length: 45
< 
* Connection #0 to host 0.0.0.0 left intact
{"maxAge":"0","toDisk":"false","maxSize":"0"}
```

The options I set are invalid, and the ContainerJFR logs do reflect that, but the response sent is still a `200 OK`. The Customizer needs refactoring to throw exceptions when its validations fail instead of simply logging them, however, that refactor should probably wait until the Command equivalent to this handler is removed so that the Customizer doesn't need to support both cases.

So, for now, please add some basic validations for these three attributes, and if the validation fails respond with some `4xx` status code.

Reply:
Here's the backend log snippet from the invalid request above:

```
[WARN] java.lang.NumberFormatException: For input string: "no"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Long.parseLong(Long.java:692)
	at java.base/java.lang.Long.parseLong(Long.java:817)
	at com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer$OptionKey$1.acceptThrows(RecordingOptionsCustomizer.java:88)
	at com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer$OptionKey$1.acceptThrows(RecordingOptionsCustomizer.java:85)
	at com.redhat.rhjmc.containerjfr.core.util.CheckedConsumer.accept(CheckedConsumer.java:49)
	at com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer.lambda$apply$0(RecordingOptionsCustomizer.java:67)
	at java.base/java.util.HashMap$Values.forEach(HashMap.java:976)
	at com.redhat.rhjmc.containerjfr.core.RecordingOptionsCustomizer.apply(RecordingOptionsCustomizer.java:67)
	at com.redhat.rhjmc.containerjfr.commands.internal.CommandsInternalModule.lambda$provideRecordingOptionsBuilderFactory$0(CommandsInternalModule.java:172)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.TargetRecordingOptionsPatchHandler.lambda$handleAuthenticated$2(TargetRecordingOptionsPatchHandler.java:117)
	at com.redhat.rhjmc.containerjfr.net.TargetConnectionManager.executeConnectedTask(TargetConnectionManager.java:86)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.TargetRecordingOptionsPatchHandler.handleAuthenticated(TargetRecordingOptionsPatchHandler.java:98)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:82)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:61)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.doEnd(BodyHandlerImpl.java:296)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.end(BodyHandlerImpl.java:276)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl.lambda$handle$0(BodyHandlerImpl.java:87)
	at io.vertx.core.http.impl.HttpServerRequestImpl.onEnd(HttpServerRequestImpl.java:521)
	at io.vertx.core.http.impl.HttpServerRequestImpl.handleEnd(HttpServerRequestImpl.java:507)
	at io.vertx.core.http.impl.Http1xServerConnection.handleEnd(Http1xServerConnection.java:176)
	at io.vertx.core.http.impl.Http1xServerConnection.handleContent(Http1xServerConnection.java:163)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:140)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:101)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1518)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1267)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1314)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
Should none of the options be set if any of them are invalid?

Reply:
Yea, that's probably best. Otherwise it will be difficult to communicate to a client that their request has been only partially accepted.

Reply:
Ok!

Reply:
My solution was very unelegant, if you can think of a more compact way I can redo it!